### PR TITLE
Extend compiler to handle lab 6

### PR DESCRIPTION
**Pull request by pavpanchekha**
This PR extends the compiler to support lab 6 and beyond. This requires a few minor changes (new functions, support for `try` and `with`, and so on) but there's a reasonably big change in that ch6 is the first chapter to use an external file (the browser stylesheet). To support this there's a new `filesystem` global which stores filename-contents pairs. The compiler translates Python's `open` function to use the `filesystem` global, and also embeds external files into the JavaScript during compile time.

### CODE REVIEW

**Review by chrishtr — COMMENTED** [review]
I think it's time to add some unittests - just comparing python in with JS out. Are you up for that? I can help out if you need it.

**Comment by chrishtr** [comment]
It looks like you accidentally included parts of PR 141?

**Comment by pavpanchekha** [comment]
Yeah, I figured that was the easiest way to test things. It sequences the PRs, which isn't great, but I hope we'll get both merged relatively soon.

**Comment by pavpanchekha** [comment]
Unit tests are in.

File: Makefile [hunk 1]
```diff
@@ -12,6 +12,7 @@ widgets: $(patsubst %,www/widgets/%,$(WIDGET_LAB_CODE))
 
 lint: book/*.md src/*.py
 	python3 infra/compare.py --config config.json
+	python3 -m doctest infra/compiler.txt
```
Comment (chrishtr) [inline Makefile:15]: compiler.md

File: Makefile [hunk 2]
```diff
@@ -27,8 +28,13 @@ www/draft/%.html: book/%.md infra/template.html infra/signup.html infra/filter.l
...
 
-www/widgets/%.js: src/%.py
-	python3 infra/compile.py $< $@ --hints src/$*.hints
+www/widgets/lab%.js: src/lab%.py src/lab%.hints infra/compile.py
+	python3 infra/compile.py $< $@ --hints src/lab$*.hints
+
+# This is not quite right in that each widget goes to the graphics chapter.
+# Unfortunately I don't have a fix for this ATM
+www/widgets/lab%-browser.html: infra/labN-browser.html
+	pandoc --variable chapter=$* --variable name=graphics --template $< book/graphics.md -o $@
```
Comment (chrishtr) [inline Makefile:37]: fixed?

File: src/lab6.hints [hunk 3]
```diff
@@ -0,0 +1,12 @@
+[
+{"line":  25, "code": "':' in host", "type": "str"},
```
Comment (chrishtr) [inline src/lab6.hints:2]: Follow-up task: remove line number feature

File: www/widgets/lab3-browser.html [hunk 4]
```diff
@@ -0,0 +1,36 @@
+<!doctype html>
```
Comment (chrishtr) [inline www/widgets/lab3-browser.html:1]: Remove obsolete files?

**Comment by pavpanchekha** [comment]
Line numbers are now optional in all hints files (and I've removed them from existing hints files) and the obsolete files are gone. I think this is ready to merge.

File: Makefile [hunk 5]
```diff
@@ -27,8 +28,11 @@ www/draft/%.html: book/%.md infra/template.html infra/signup.html infra/filter.l
 www/rss.xml: news.yaml infra/rss-template.xml
 	pandoc --template infra/rss-template.xml  -f markdown -t html $< -o $@
 
-www/widgets/%.js: src/%.py
-	python3 infra/compile.py $< $@ --hints src/$*.hints
+www/widgets/lab%.js: src/lab%.py src/lab%.hints infra/compile.py
+	python3 infra/compile.py $< $@ --hints src/lab$*.hints
+
+www/widgets/lab%-browser.html: infra/labN-browser.html infra/labN-browser.lua config.json
+	pandoc --lua-filter=infra/labN-browser.lua --metadata-file=config.json --metadata chapter=$* --template $< book/graphics.md -o $@
```
Comment (chrishtr) [inline Makefile:35]: graphics.md still here?

**Review by chrishtr — APPROVED** [review]
One last comment. Ok to merge after resolved.

File: Makefile (hunk 5 above)
Comment (pavpanchekha) [reply Makefile:35]: Hahaha oh boy you won't like this but... `pandoc` needs an input file. Literally any file will do—`/dev/null` is fine actually but then there are warnings (what file type is null?!). I'll change it to `index.md` I guess.

**Pull request merged**